File: frontend/src/types/alerts.ts
export type AlertFormat = 'cef' | 'syslog_3164' | 'syslog_5424' | 'json'
export type AlertSeverity = 'LOW' | 'MEDIUM' | 'HIGH' | 'CRITICAL'

export interface GeoInfo {
  country_code: string
  country_name: string
  city: string | null
  asn: string | null
  is_known_vpn: boolean
}

export interface ThreatIntelInfo {
  reputation_score: number
  is_known_bad: boolean
  categories: string[]
  source: string
}

export interface AssetInfo {
  hostname: string
  tier: 'CROWN_JEWEL' | 'STANDARD' | 'LOW' | 'UNKNOWN'
  owner_team: string | null
  os: string | null
}

export interface NormalizedEvent {
  id: string
  timestamp: string
  source_format: AlertFormat
  severity: AlertSeverity
  event_type: string
  source_ip: string | null
  dest_ip: string | null
  source_host: string | null
  dest_host: string | null
  src_port: number | null
  dest_port: number | null
  username: string | null
  process_name: string | null
  message: string
  raw_fields: Record<string, unknown>
  geo: GeoInfo | null
  threat_intel: ThreatIntelInfo | null
  asset: AssetInfo | null
  ingested_at: string
}

export interface RawAlertIn {
  raw: string
  format_hint?: AlertFormat
}

File: backend/app/models/alerts.py
from datetime import datetime, timezone
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field

from app.models.enrichment import AssetInfo, GeoInfo, ThreatIntelInfo


class AlertFormat(StrEnum):
    CEF = "cef"
    SYSLOG_3164 = "syslog_3164"
    SYSLOG_5424 = "syslog_5424"
    JSON = "json"


class AlertSeverity(StrEnum):
    LOW = "LOW"
    MEDIUM = "MEDIUM"
    HIGH = "HIGH"
    CRITICAL = "CRITICAL"


class RawAlertIn(BaseModel):
    raw: str = Field(..., max_length=65_536, description="Raw alert string in any supported format")
    format_hint: AlertFormat | None = Field(
        None, description="Optional format hint; auto-detected if omitted"
    )


class NormalizedEvent(BaseModel):
    id: str
    timestamp: datetime
    source_format: AlertFormat
    severity: AlertSeverity
    event_type: str
    source_ip: str | None = None
    dest_ip: str | None = None
    source_host: str | None = None
    dest_host: str | None = None
    src_port: int | None = None
    dest_port: int | None = None
    username: str | None = None
    process_name: str | None = None
    message: str
    raw_fields: dict[str, Any] = Field(default_factory=dict)
    geo: GeoInfo | None = None
    threat_intel: ThreatIntelInfo | None = None
    asset: AssetInfo | None = None
    ingested_at: datetime = Field(default_factory=lambda: datetime.now(tz=timezone.utc))
